python_modules/dagster-graphql/dagster_graphql/client/client.py

```diff
@@ -396,3 +400,189 @@ def terminate_run(self, run_id: str):
+        """
+        check.list_param(metrics, "metrics", of_type=Metric)
+        check.inst_param(context, "context", OpExecutionContext)
+        metric_graphql_input = {}
+
+        if context.dagster_run.external_job_origin is None:
+            raise Exception("dagster run for this context has not started yet")
+
+        if context.has_assets_def:
+            metric_graphql_input = {
+                "assetMetricDefinitions": [
+                    {
+                        "runId": context.run_id,
+                        "stepKey": context.get_step_execution_context().step.key,
+                        "codeLocationName": context.dagster_run.external_job_origin.location_name,
+                        "repositoryName": (
+                            context.dagster_run.external_job_origin.external_repository_origin.repository_name
+                        ),
+                        "assetKey": selected_asset_key,
+                        "assetGroup": context.assets_def.group_names_by_key.get(
+                            selected_asset_key, None
+                        ),
+                        "metricValues": [
+                            {
+                                "metricValue": metric_def.metric_value,
+                                "metricName": metric_def.metric_name,
+                            }
+                            for metric_def in metrics
+                        ],
+                    }
+                    for selected_asset_key in context.selected_asset_keys
+                ]
+            }
+        else:
+            metric_graphql_input = {
+                "jobMetricDefinitions": [
+                    {
+                        "runId": context.run_id,
+                        "stepKey": context.get_step_execution_context().step.key,
+                        "codeLocationName": context.dagster_run.external_job_origin.location_name,
+                        "repositoryName": (
+                            context.dagster_run.external_job_origin.external_repository_origin.repository_name
+                        ),
+                        "metricValues": [
+                            {
+                                "metricValue": metric_def.metric_value,
+                                "metricName": metric_def.metric_name,
+                            }
+                            for metric_def in metrics
+                        ],
+                    }
+                ]
+            }
+
+        res_data: Dict[str, Dict[str, Any]] = self._execute(
+            PUT_CLOUD_METRICS_MUTATION, {"metrics": [metric_graphql_input]}
+        )
+
+        store_metric_result: Dict[str, Any] = res_data["createExternalMetrics"]
+        store_metric_result_type: str = store_metric_result["__typename"]
+        if store_metric_result_type == "CreateExternalMetricsSuccess":
+            return
+
+        raise DagsterGraphQLClientError(store_metric_result_type, store_metric_result["message"])
```

Comment:
This is a bit tricky, we can address this in our examples I guess, but there's a strange case for when an op/asset succeeds but the metric submission fails. In that case we might want to do some sort of soft failure without actually marking the execution as failed?

This problem isn't unique to metrics (doing work after actual materialization/body of an op), but w/ this API we are actively encouraging this pattern.